src/MongoDB.Bson/EqualsPoc/EquatableBaseClass.cs

```
@@ -0,0 +1,32 @@
+﻿using System;
+
+namespace MongoDB.Bson.EqualsPoc
+{
+    internal class EquatableBaseClass : IEquatable<EquatableBaseClass>
+    {
+        public object _ref1;
+        public int _value1;
+
+        public EquatableBaseClass(object ref1, int value1)
+        {
+            _ref1 = ref1;
+            _value1 = value1;
+        }
+
+        public override bool Equals(object obj) =>
+            Equals(obj as EquatableBaseClass);
+
+        public bool Equals(EquatableBaseClass other) =>
+            object.ReferenceEquals(this, other) ||
```

I'd suggest to check for `null` first.
I think in general a comparison to a different non-derived type is more likely than to the same instance. 
This will also fail faster for `null`-case for derived classes (instead of two subsequent `object.ReferenceEquals(this, other)` calls)
Maybe this is one of the reasons that null check is first in MS examples.

---

Good suggestions.

I've refactored to compare to null first.